add docs for AWS assumable identities (#1179)
chainguard-dev/edu: content/chainguard/chainguard-enforce/authentication/identity-examples/enforce-aws-identity.md

We will be using Terraform to create an identity for an AWS role to assume. This step outlines how to create the Terraform configuration files that, together, will produce such an identity.

These files are available here: https://github.com/chainguard-dev/platform-examples/tree/main/aws-auth

To help explain each configuration file's purpose, we will go over what they do one by one. First, though, create a directory to hold the Terraform configuration and navigate into it.

```sh
mkdir ~/enforce-aws && cd $_
```

This will help make it easier to clean up your system at the end of this guide.

### `main.tf`

The first file, which we will call `main.tf`, will serve as the scaffolding for our Terraform infrastructure.

The file will consist of the following content.

```
terraform {
  required_providers {
    aws        = { source = "hashicorp/aws" }
    chainguard = { source = "chainguard-dev/chainguard" }
    ko         = { source = "ko-build/ko" }
  }
}
```

This is a fairly barebones Terraform configuration file, but we will define the rest of the resources in the other two files. In `main.tf`, we declare and initialize the Chainguard Terraform provider.

> Next, you can look at the `chainguard.tf` file.

Reviewer: ```suggestion
Next, you can create the `chainguard.tf` file.
```

It's a bit pedantic but I try to avoid words like "look" or "see" in documentation to make it more inclusive for visually impaired folks

> ### `chainguard.tf`

Reviewer: I think this section needs another look, it's missing quite a bit from the chainguard.tf file on the example repo: https://github.com/chainguard-dev/platform-examples/blob/main/aws-auth/iac/chainguard.tf

`chainguard.tf` will create a couple of structures that will help us test out the identity in a workflow.

This Terraform configuration consists of two main parts. The first part of the file will contain the following lines.

```
resource "chainguard_group" "user-group" {
  name   	 = "user-group"
  description = <<EOF
    This group simulates an end-user group, which the AWS role identity
    can interact with via the identity in aws.tf.
  EOF
}
```

This section creates a Chainguard IAM group named `example-group`, as well as a description of the group. This will serve as some data for the identity — which will be created by the `lambda.tf` file — to access when we test it out later on.

> Now you can move on to creating the last of our Terraform configuration files, `aws.tf`.

Reviewer: ```suggestion
Now you can move on to creating the last of our Terraform configuration files, `lambda.tf`.
```

The `aws_user_id_pattern` field configures the identity to be assumable only by the AWS role with the specified name, which is itself assumed by another execution role, which we'll configure below. This role will be assumed so can't use the role ARN directly in `aws_arn`; We must used the ARN of the assumed role.

The section after that looks up the `viewer` role.

```
data "chainguard_roles" "viewer" {
  name = "viewer"
}
```

The final section grants this role to the identity on the `example-group`.

```
resource "chainguard_rolebinding" "view-stuff" {
  identity = chainguard_identity.aws.id
  group	= chainguard_group.example-group.id
  role 	= data.chainguard_roles.viewer.items[0].id
}
```

After defining these resources, there are some other resources in the example directory that build and deploy a Lambda function that assumes the identity. We'll describe that code in the next section.

After defining these resources, your Terraform configuration will be ready. Now you can run a few `terraform` commands to create the resources defined in your `.tf` files.

> ### `lambda.tf`

Reviewer: I think things got a little twisted here - we can delete this whole last section, right?

> The file file, `lambda.tf`, will create an AWS Lambda function that will assume the identity you created in the previous section. This function will then use the identity to interact with Chainguard resources.

Reviewer: ```suggestion
The `lambda.tf` file will create an AWS Lambda function that will assume the identity you created in the previous section. This function will then use the identity to interact with Chainguard resources.
```

This section defines an IAM policy document that will allow the AWS Lambda service to assume the role you'll create in the next section:

```hcl
data "aws_iam_policy_document" "assume_role" {
  statement {
    effect = "Allow"

    principals {
      type        = "Service"
      identifiers = ["lambda.amazonaws.com"]
    }

    actions = ["sts:AssumeRole"]
  }
}
```

This section defines the role that the AWS Lambda service will assume:

```hcl
resource "aws_iam_role" "iam_for_lambda" {
  name               = "iam_for_lambda"
  assume_role_policy = data.aws_iam_policy_document.assume_role.json
}
```

The final section defines a AWS lambda function implemented in Go that will assume the identity you created in the previous section:

```hcl
resource "aws_lambda_function" "test_lambda" {
  filename      = "lambda_function_payload.zip"
  function_name = "lambda_function_name"
  role          = aws_iam_role.iam_for_lambda.arn
  handler       = "bootstra"

  source_code_hash = data.archive_file.lambda.output_base64sha256

  runtime = "go1.x"

  environment {
    variables = {
      foo = "bar"
    }
  }
}
```

> See [this basic example](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_function#basic-example) for configuring AWS Lambda using Terraform, and the docs for [deploying Go Lambda functions with .zip file archives](https://docs.aws.amazon.com/lambda/latest/dg/golang-package.html) for more information on how to configure the `filename` and `source_code_hash` fields.

Reviewer: ```suggestion
Check out [this basic example](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_function#basic-example) for configuring AWS Lambda using Terraform, and the docs for [deploying Go Lambda functions with .zip file archives](https://docs.aws.amazon.com/lambda/latest/dg/golang-package.html) for more information on how to configure the `filename` and `source_code_hash` fields.
```

## Creating Your Resources

First, run `terraform init` to initialize Terraform's working directory.

```sh
terraform init
```

Then run `terraform plan`. This will produce a speculative execution plan that outlines what steps Terraform will take to create the resources defined in the files you set up in the last section.

```sh
terraform plan
```

Then apply the configuration.

```sh
terraform apply
```

Before going through with applying the Terraform configuration, this command will prompt you to confirm that you want it to do so. Enter `yes` to apply the configuration.

```
...

Plan: 8 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + aws-identity = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value:
```

After typing `yes` and pressing `ENTER`, the command will complete and will output an `aws-identity` value.

```
...

Apply complete! Resources: 8 added, 0 changed, 0 destroyed.

Outputs:

aws-identity = "<your identity>"
```

> This is the identity's [UIDP (unique identity path)](/chainguard/chainguard-enforce/reference/events/#uidp-identifiers), which you configured the `aws.tf` file to emit in the previous section. Note this value down, as you'll need it to set up the AWS role you'll use to test the identity. If you need to retrieve this UIDP later on, though, you can always run the following `chainctl` command to obtain a list of the UIDPs of all your existing identities.

Reviewer: ```suggestion
This is the identity's [UIDP (unique identity path)](/chainguard/chainguard-enforce/reference/events/#uidp-identifiers), which you configured the `chainguard.tf` file to emit in the previous section. Note this value down, as you'll need it to set up the AWS role you'll use to test the identity. If you need to retrieve this UIDP later on, though, you can always run the following `chainctl` command to obtain a list of the UIDPs of all your existing identities.
```

I think?

This can be used to authorize requests from AWS Lambda, ECS, EKS, or any other AWS service that supports [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html).

## Prerequisites

To complete this guide, you will need the following.

* `terraform` installed on your local machine. Terraform is an open-source Infrastructure as Code tool which this guide will use to create various cloud resources. Follow [the official Terraform documentation](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli) for instructions on installing the tool.
* `chainctl` — the Chainguard Enforce command line interface tool — installed on your local machine. Follow our guide on [How to Install `chainctl`](/chainguard/chainguard-enforce/how-to-install-chainctl/) to set this up.
* An AWS account with the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html) installed and configured. The [Terraform provider for AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) uses credentials configured via the AWS CLI.

> 

Reviewer: The testing section assumes readers have Go - could you add a bullet to this section that specifies that?

## Creating Terraform Files

> We will be using Terraform to create an identity for an AWS role to assume. This step outlines how to create three Terraform configuration files that, together, will produce such an identity.

Reviewer: ```suggestion
We will be using Terraform to create an identity for an AWS role to assume. This step outlines how to create four Terraform configuration files that, together, will produce such an identity.
```

  document = jsonencode({
    apiVersion = "policy.sigstore.dev/v1beta1"
    kind  	 = "ClusterImagePolicy"
    metadata = {
      name = "trust-any-cgr"
    }
    spec = {
      images = [{
        glob = "cgr.dev/**"
      }]
      authorities = [{
        static = {
          action = "pass"
        }
      }]
    }
  })
}
```

This policy trusts everything coming from the [Chainguard Registry](/chainguard/chainguard-images/registry/overview/). Because this policy is broadly permissive, it wouldn't be practical or secure to use in a real-world scenario. Like the example group, this policy serves as some data for the AWS role to inspect after it assumes the Chainguard identity.

Now you can move on to creating the last of our Terraform configuration files, `aws.tf`.

### `aws.tf`

The `aws.tf` file is what will actually create the identity for your AWS role to assume. The file will consist of four sections, which we'll go over one by one.

The first section creates the identity itself.

```
resource "chainguard_identity" "aws" {
  parent_id   = var.group
  name        = "aws-identity"
  description = "Identity for AWS role to assume"

>   aws_identity {

Reviewer: so full disclosure: I am a total AWS noob and it's been a while since I played around with it. 

having said that, I'm unclear about some of the details i'm supposed to grab from my account for this file. `aws_account` is clear enough but I'm not sure what `aws_user_id_pattern` is supposed to look like. In the text below this block where you explain its contents, could you elaborate a bit on how a reader could obtain this? Maybe provide a dummy example of what this field might look like?

    aws_account         = "[AWS-ACCOUNT-ID]"
    aws_user_id_pattern = "^AROA(.*):[AWS-ROLE-SESSION-NAME]$"

>     // NB: This role will be assumed so can't use the role ARN directly. We must used the ARN of the assumed role

Reviewer: Could you move this detail to the text below? A lot of readers (including me) tend to ignore comments in code blocks like this

The most important part of this section is the `aws_identity`. When the AWS role tries to assume this identity later on, it must present a token matching the `aws_account`, `aws_user_id_pattern`, and `aws_arn` specified here in order to do so.

The next section will output the new identity's `id` value. This is a unique value that represents the identity itself.

```
output "aws-identity" {
  value = chainguard_identity.aws.id
}
```

The section after that looks up the `viewer` role.

```
data "chainguard_roles" "viewer" {
  name = "viewer"
}
```

The final section grants this role to the identity on the `example-group`.

```
resource "chainguard_rolebinding" "view-stuff" {
  identity = chainguard_identity.aws.id
  group	= chainguard_group.user-group.id
  role 	= data.chainguard_roles.viewer.items[0].id
}
```

After defining these resources, your Terraform configuration will be ready. Now you can run a few `terraform` commands to create the resources defined in your `.tf` files.

## Using the identity with AWS Lambda

> In this section, you'll create an AWS Lambda function that will assume the identity you created in the previous section. This function will then use the identity to interact with Chainguard resources.

Reviewer: ```suggestion
The final file, `lambda.tf`, will create an AWS Lambda function that will assume the identity you created in the previous section. This function will then use the identity to interact with Chainguard resources. This file will consist of three sections.
```

> First, define the following resources in a `lambda.tf` file. The file will consist of three sections, which we'll go over one by one.

Reviewer: ```suggestion

```

This section defines an IAM policy document that will allow the AWS Lambda service to assume the role you'll create in the next section:

```hcl
data "aws_iam_policy_document" "assume_role" {
  statement {
    effect = "Allow"

    principals {
      type        = "Service"
      identifiers = ["lambda.amazonaws.com"]
    }

    actions = ["sts:AssumeRole"]
  }
}
```

This section defines the role that the AWS Lambda service will assume:

```hcl
resource "aws_iam_role" "iam_for_lambda" {
  name               = "iam_for_lambda"
  assume_role_policy = data.aws_iam_policy_document.assume_role.json
}
```

> This section defines a AWS lambda function implemented in Node.js that will assume the identity you created in the previous section:

Reviewer: ```suggestion
The final section defines a AWS lambda function implemented in Node.js that will assume the identity you created in the previous section:
```

I like leading phrases like this, it lets the reader know the end is in sight

```hcl
resource "aws_lambda_function" "test_lambda" {
  filename      = "lambda_function_payload.zip"
  function_name = "lambda_function_name"
  role          = aws_iam_role.iam_for_lambda.arn
  handler       = "index.test"

>   source_code_hash = data.archive_file.lambda.output_base64sha256

Author: I trimmed out a bit of stuff for brevity, but I can link to where it's described more in detail in TF and AWS docs.

I wanted to avoid making this a "how to deploy a Go Lambda function" doc, since many readers may not be doing that, it's just necessary to test the identity.

  runtime = "nodejs18.x"

  environment {
    variables = {
      foo = "bar"
    }
  }
}
```

> When this function is invoked, it will assume the AWS role you created in the previous section, and be able to present credentials as that AWS role that will allow it to assume the Chainguard identity you created in the previous section, to view and manage Chainguard resources.

Reviewer: ```suggestion
When this function is invoked, it will assume the AWS role you created in the previous section. It will then be able to present credentials as that AWS role that will allow it to assume the Chainguard identity you created in the previous section, to view and manage Chainguard resources.
```

## Creating Your Resources

First, run `terraform init` to initialize Terraform's working directory.

```sh
terraform init
```

Then run `terraform plan`. This will produce a speculative execution plan that outlines what steps Terraform will take to create the resources defined in the files you set up in the last section.

```sh

> terraform plan

Reviewer: ```suggestion
terraform plan
```
Again, i could not get past this command successfully. I'll still review the rest of the content but I'll need to pass a tech test before I can approve it.

...

Plan: 8 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + aws-identity = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value:
```

After typing `yes` and pressing `ENTER`, the command will complete and will output an `aws-identity` value.

```
...

Apply complete! Resources: 8 added, 0 changed, 0 destroyed.

Outputs:

aws-identity = "<your identity>"
```

This is the identity's [UIDP (unique identity path)](/chainguard/chainguard-enforce/reference/events/#uidp-identifiers), which you configured the `aws.tf` file to emit in the previous section. Note this value down, as you'll need it to set up the AWS role you'll use to test the identity. If you need to retrieve this UIDP later on, though, you can always run the following `chainctl` command to obtain a list of the UIDPs of all your existing identities.

```sh
chainctl iam identities ls
```

### Testing the Identity

When the AWS Lambda function is invoked, first it needs to get its credentials, which assert that it is the AWS IAM role defined earlier.

In Go, you can do this with `aws-sdk-go-v2`:

> ```go

Reviewer: Again, I didn't actually get this far but I can tell I'm going to have a hard time with this test as it's written. Could you make this a bit more idiot-proof? Or Mark-proof if that's a less problematic way to put it. 

Basically, I'd wrap this Go code in a `cat` command to create the files and then add whatever command a reader should run to invoke the code. 

I try to always give readers every command they need to perform to run code or whatever. 

Can you make it a bit more idiot-proof?
